### DISCUSSION TITLE
remove hardcoded istio:ingressgateway selector for ingress gateway

### DISCUSSION THREAD

**Discussion by jrhunger** [kiali/kiali#7226] [Q&A]
### What do you want to improve?
Allow the istio ingress gateway selector to be configurable, rather than hardcoding to istio: ingressgateway

### What is the current behavior?
The Istio Gateway helm chart [General Concerns](https://artifacthub.io/packages/helm/istio-official/gateway#general-concerns) section mentions that the selector will be set to the release name minus istio- (e.g. ingress for istio-ingress, or gateway for istio-gateway).  So there is expectation on Istio side that this is not always istio: ingressgateway

However if anything other than istio:ingressgateway is used, Kiali doesn't find the ingress gateway and presents a red error "istio-ingressgateway Not Found"

### What is the new behavior?
Istio Ingress Gateway selector can be configured by a configMap entry or whatever else makes more sense.

**Comment by jmazzitelli** [comment]
> Allow the istio ingress gateway selector to be configurable, rather than hardcoding to istio: ingressgateway

Where is this hardcoded within the Kiali codebase?

Unless you are referring to something else, I believe this is already configurable.

> if anything other than istio:ingressgateway is used, Kiali doesn't find the ingress gateway and presents a red error "istio-ingressgateway Not Found"

Have you looked at the kiali.io docs regarding customizing the component status fields for this kind of thing?

Please see:
* https://kiali.io/docs/faq/istio-component-status/
* https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.istio.component_status.components[*].app_label

**Comment by jmazzitelli** [comment]
(_I moved this to Q&A Discussions -- I believe this is already configurable and documented on kiali.io documentation website_)

**Reply by jrhunger** [reply]
Here are some places i found it hardcoded.  

This one may be just a default that can get overridden later?
[GatewayForm.tsx](https://github.com/kiali/kiali/blob/d91b83bac66f9cd5f82ac7491ae34f9d8816c991/frontend/src/pages/IstioConfigNew/GatewayForm.tsx#L29)
```
export const initGateway = (): GatewayState => ({
  addWorkloadSelector: false,
  gatewayServers: [],
  serversForm: [],
  workloadSelectorLabels: 'istio=ingressgateway',
  workloadSelectorValid: true
});
```

[workload.go:isGateway()](https://github.com/kiali/kiali/blob/d91b83bac66f9cd5f82ac7491ae34f9d8816c991/models/workload.go#L442)
```
		if labelValue, ok := workload.Labels["istio"]; ok && (labelValue == "ingressgateway" || labelValue == "egressgateway") {
			return true
		}
```

But also i had not seen that part of the docs and did not realize how extensive the CRD can be.  I will play around with those settings and see if I can get it to work properly.

**Reply by jmazzitelli** [reply]
@hhovsepy can you take a look at those two pieces of frontend code that @jrhunger mentioned above? Those hardcoded values do not look right to me - but I don't know this code at all.

**Reply by hhovsepy** [reply]
@jrhunger can you please provide a screenshot or a steps where red error "istio-ingressgateway Not Found" happens?
In the first code example, yes, this is a default value in UI input form which can be overridden.
The second example, it is wrongly hardcoded, it will be fixed to read from config. 
But I believe these are not related to the red error message you see.
The error message can happen when "istio-ingressgateway" is located in different namespace, but it can be configured as well https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.istio.component_status.components[*].namespace.

**Reply by jrhunger** [reply]
Here is where i see it.
![kiali-istio-components1](https://github.com/kiali/kiali/assets/7697261/23b693b3-ca71-4b36-abea-f2fdf775bbb9)
Looking at my test traffic, i see the istio ingress as a component.  I'm still trying to figure out what i should see here (and as you can see doing some knative testing at the same time)
![image](https://github.com/kiali/kiali/assets/7697261/f91e7527-9ad1-45b7-88c4-3faa21e2b9e5)

**Reply by hhovsepy** [reply]
Thanks @jrhunger , so this is a configuration issue.
You need to properly set `app_label` and `namespace` parameters for "ingress" component in a configuration https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.istio.component_status.components.
The hardcoded values will be fixed in a separate issue https://github.com/kiali/kiali/issues/7232.
Thanks for reporting.

**Reply by jrhunger** [reply]
Thank you. I added this to my kiali CR and the Istio Components Status bullseye went away!
```
  external_services:
    istio:
      component_status:
        components:
          - app_label: "istio-ingress"
            namespace: "istio-ingress"
            is_core: true
            is_proxy: true
    grafana:
      enabled: false
```
